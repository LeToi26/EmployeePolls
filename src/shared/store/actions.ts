import { InitialState, Questions, User, Users } from "shared/model/poll";
import productApi from "../api/productApi";
import { Dispatch } from "@reduxjs/toolkit";

export const RECEIVE_USERS = "RECEIVE_USERS";
export const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";

export const SET_CURRENT_USER = "SET_CURRENT_USER";
export const LOGOUT_CURRENT_USER = "LOGOUT_CURRENT_USER";

export const CREATE_QUESTION = "CREATE_QUESTION";
export const ADD_ANSWER_QUESTION = "ADD_ANSWER_QUESTION";

export const ADD_ANSWER_USER = "ADD_ANSWER_USER";
export const CREATE_QUESTION_USER = "CREATE_QUESTION_USER";

export function initialData() {
  return (dispatch: Dispatch) => {
    return productApi
      .getInitialData()
      .then(({ users, questions }: { users: any; questions: any }) => {
        dispatch(receiveUsers(users));
        dispatch(receiveQuestions(questions));
      });
  };
}

export function receiveUsers(users: Users) {
  return {
    type: RECEIVE_USERS,
    users,
  };
}

export function receiveQuestions(questions: Questions) {
  return {
    type: RECEIVE_QUESTIONS,
    questions,
  };
}

export function handleLogout() {
  return (dispatch: Dispatch) => {
    return dispatch(logoutCurrentUser());
  };
}

export function logoutCurrentUser() {
  return {
    type: LOGOUT_CURRENT_USER,
  };
}

export function handleLogin(username: string, password: string) {
  return (dispatch: Dispatch, getState: () => { State: InitialState }) => {
    const { State } = getState();
    const user: any = Object.values(State.users).find(
      (user: any) => user.id === username && user.password === password
    );
    if (!!user) {
      return dispatch(setCurrentUser(user));
    }
  };
}

export function setCurrentUser(authedUser: User | {}) {
  return {
    type: SET_CURRENT_USER,
    authedUser,
  };
}

export function handleAddAnswer(questionId: string, answer: string) {
  return (dispatch: Dispatch, getState: () => { State: InitialState }) => {
    const { State } = getState();
    return productApi
      .saveQuestionAnswer(State.authedUser.id, questionId, answer)
      .then(() => {
        dispatch(addAnswerQuestion(State.authedUser.id, questionId, answer));
        dispatch(addAnswerUser(State.authedUser.id, questionId, answer));
      });
  };
}

function addAnswerQuestion(author: string, qid: string, answer: string) {
  return {
    type: ADD_ANSWER_QUESTION,
    author,
    qid,
    answer,
  };
}

export function addAnswerUser(authedUser: string, qid: string, answer: string) {
  return {
    type: ADD_ANSWER_USER,
    authedUser,
    qid,
    answer,
  };
}

export function handleCreateQuestion(
  firstOption: string,
  secondOption: string
) {
  return (dispatch: Dispatch, getState: () => { State: InitialState }) => {
    const { State } = getState();

    return productApi
      .saveQuestion(firstOption, secondOption, State.authedUser)
      .then((question) => {
        dispatch(createQuestion(question));
        dispatch(createQuestionUser(question));
      });
  };
}

export function createQuestionUser(question: any) {
  const { author, id } = question;
  return {
    type: CREATE_QUESTION_USER,
    author,
    qid: id,
  };
}

function createQuestion(question: any) {
  return {
    type: CREATE_QUESTION,
    question,
  };
}
