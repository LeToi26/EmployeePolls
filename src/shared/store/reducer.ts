import { InitialState } from "shared/model/poll";
import {
  SET_CURRENT_USER,
  LOGOUT_CURRENT_USER,
  RECEIVE_USERS,
  RECEIVE_QUESTIONS,
  CREATE_QUESTION,
  CREATE_QUESTION_USER,
  ADD_ANSWER_USER,
  ADD_ANSWER_QUESTION,
} from "./actions";

const initialState: InitialState = {
  authedUser: null,
  users: {},
  questions: {},
};

export default function State(state = initialState, action: any) {
  switch (action.type) {
    case SET_CURRENT_USER:
      return {
        ...state,
        authedUser: action.authedUser,
      };
    case LOGOUT_CURRENT_USER:
      return {
        ...state,
        authedUser: null,
      };
    case RECEIVE_USERS:
      return {
        ...state,
        users: action.users,
      };
    case RECEIVE_QUESTIONS:
      return {
        ...state,
        questions: action.questions,
      };
    case ADD_ANSWER_USER:
      return {
        ...state,
        users: {
          ...state.users,
          [action.authedUser]: {
            ...state.users[action.authedUser as keyof typeof state],
            answers: {
              ...state.users[action.authedUser as keyof typeof state].answers,
              [action.qid]: action.answer,
            },
          },
        },
      };
    case ADD_ANSWER_QUESTION:
      return {
        ...state,
        questions: {
          ...state.questions,
          [action.qid]: {
            ...state.questions[action.qid as keyof typeof state],
            [action.answer]: {
              ...state.questions[action.qid as keyof typeof state][
                action.answer
              ],
              votes: state.questions[action.qid as keyof typeof state][
                action.answer
              ].votes.concat(action.author),
            },
          },
        },
      };
    case CREATE_QUESTION_USER:
      return {
        ...state,
        users: {
          ...state.users,
          [action.author]: {
            ...state.users[action.author as keyof typeof state],
            questions: state.users[
              action.author as keyof typeof state
            ].questions.concat(action.qid),
          },
        },
      };
    case CREATE_QUESTION:
      return {
        ...state,
        questions: {
          ...state.questions,
          [action.question.id]: action.question,
        },
      };
    default:
      return state;
  }
}
