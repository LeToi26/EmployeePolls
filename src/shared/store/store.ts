import { configureStore } from "@reduxjs/toolkit";
import State from "./reducer";

export const store = configureStore({
  reducer: {
    State,
  },
});
