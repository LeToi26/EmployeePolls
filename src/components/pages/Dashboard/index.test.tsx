/* eslint-disable testing-library/prefer-screen-queries */
/* eslint-disable testing-library/render-result-naming-convention */

import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from "../../../shared/store/store";
import { BrowserRouter } from "react-router-dom";
import DashBoard from "./index";

describe("DashBoard", () => {
  it("should render the component", () => {
    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <DashBoard />
        </BrowserRouter>
      </Provider>
    );
    expect(component).toBeDefined();
    expect(component).toMatchSnapshot();
  });
});
