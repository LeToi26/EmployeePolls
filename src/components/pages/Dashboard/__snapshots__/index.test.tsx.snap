// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashBoard should render the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="home"
      >
        <h1
          class="text-3xl font-bold mt-9"
          data-testid="dashboard"
        >
          Dashboard
        </h1>
        <ul
          class="mb-5 flex list-none flex-col flex-wrap pl-0 md:flex-row justify-center"
          data-te-nav-ref="true"
          id="pills-tab"
          role="tablist"
        >
          <li
            role="presentation"
          >
            <a
              aria-controls="new"
              aria-selected="true"
              class="my-2 block rounded bg-neutral-100 px-7 pb-3.5 pt-4 text-xs font-medium uppercase leading-tight text-neutral-500 data-[te-nav-active]:!bg-primary-100 data-[te-nav-active]:text-primary-700 dark:bg-neutral-700 dark:text-white dark:data-[te-nav-active]:text-primary-700 md:mr-4"
              data-te-nav-active="true"
              data-te-target="#new"
              data-te-toggle="pill"
              href="#new"
              id="new-tab"
              role="tab"
            >
              New Questions
            </a>
          </li>
          <li
            role="presentation"
          >
            <a
              aria-controls="answered"
              aria-selected="false"
              class="my-2 block rounded bg-neutral-100 px-7 pb-3.5 pt-4 text-xs font-medium uppercase leading-tight text-neutral-500 data-[te-nav-active]:!bg-primary-100 data-[te-nav-active]:text-primary-700 dark:bg-neutral-700 dark:text-white dark:data-[te-nav-active]:text-primary-700 md:mr-4"
              data-te-target="#answered"
              data-te-toggle="pill"
              href="#answered"
              id="answered-tab"
              role="tab"
            >
              Answered Questions
            </a>
          </li>
        </ul>
        <div>
          <h2
            class="text-2xl font-bold mt-6"
          >
            New Questions
          </h2>
          <ul
            class="divide-y divide-gray-100"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="home"
    >
      <h1
        class="text-3xl font-bold mt-9"
        data-testid="dashboard"
      >
        Dashboard
      </h1>
      <ul
        class="mb-5 flex list-none flex-col flex-wrap pl-0 md:flex-row justify-center"
        data-te-nav-ref="true"
        id="pills-tab"
        role="tablist"
      >
        <li
          role="presentation"
        >
          <a
            aria-controls="new"
            aria-selected="true"
            class="my-2 block rounded bg-neutral-100 px-7 pb-3.5 pt-4 text-xs font-medium uppercase leading-tight text-neutral-500 data-[te-nav-active]:!bg-primary-100 data-[te-nav-active]:text-primary-700 dark:bg-neutral-700 dark:text-white dark:data-[te-nav-active]:text-primary-700 md:mr-4"
            data-te-nav-active="true"
            data-te-target="#new"
            data-te-toggle="pill"
            href="#new"
            id="new-tab"
            role="tab"
          >
            New Questions
          </a>
        </li>
        <li
          role="presentation"
        >
          <a
            aria-controls="answered"
            aria-selected="false"
            class="my-2 block rounded bg-neutral-100 px-7 pb-3.5 pt-4 text-xs font-medium uppercase leading-tight text-neutral-500 data-[te-nav-active]:!bg-primary-100 data-[te-nav-active]:text-primary-700 dark:bg-neutral-700 dark:text-white dark:data-[te-nav-active]:text-primary-700 md:mr-4"
            data-te-target="#answered"
            data-te-toggle="pill"
            href="#answered"
            id="answered-tab"
            role="tab"
          >
            Answered Questions
          </a>
        </li>
      </ul>
      <div>
        <h2
          class="text-2xl font-bold mt-6"
        >
          New Questions
        </h2>
        <ul
          class="divide-y divide-gray-100"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
